gossip/src/cluster_info.rs

```diff
@@ -680,6 +680,25 @@ impl ClusterInfo {
         self.my_contact_info.read().unwrap().id
     }
 
+    pub fn keypair(&self) -> RwLockReadGuard<Arc<Keypair>> {
+        self.keypair.read().unwrap()
+    }
+
+    pub fn set_keypair(&self, new_keypair: Arc<Keypair>) {
```

Comment:
@behzadnouri, does `set_keypair()` look ok?  Anything else I might be missing to encourage gossip to jump over to the new keypair?

Reply:
It looks ok to me but I am bit worried to claim that nothing is left out because, well, you never know.

This seems a very nuanced change, and feels to me a bit risky. Is there any way to avoid adding this capability?

Reply:
> This seems a very nuanced change, and feels to me a bit risky. Is there any way to avoid adding this capability?

I don't think so. Basically the bigger picture here is to enable online rollover from one validator to another, for upgrades or failover purposes. One has a backup unstaked validator running with another identity, and then upon signal it needs to flip over to using the staked validator identity and start voting. The "start voting" part is not in this PR, that's next and will be a much larger change/complex change.

Reply:
Would it make sense to wrap the key-pair in a struct, such as
```
struct ValidatorIdentity(Arc<RwLock<Keypair>>)
```
and only allow keypair to be borrowed, as in:
```
impl ValidatorIdentity {
   fn keyapir(&self) -> RwLockReadGuard<Keypair> {
     self.0.read().unwrap()
   }
}
```
I think this would prevent the `Keypair` or `Arc<Keypair>` leaking elsewhere and would enforce some type-safety. e.g. `ClusterInfo` would then have a `ValidatorIdentity` field.

Reply:
🤔 I'll give that a shot and see what the code looks like. I do like the idea as a way to prevent the "swappable" keypair from escaping.

Reply:
This almost worked out.  The problem is Replay stage.  On an identity change, Replay needs to hold onto the previous identity until the next fork reset and not doing so would cause significant code churn in Replay.  

So for `ValidatorIdentity` to work out, I think it needs to look more like
```rust
struct GossipIdentity {
  keypair: Keypair,
  my_contact_info: ContactInfo,
  instance: NodeInstance,
}
```
and then `ClusterInfo` holds a `gossip_id: RwLock<Arc<GossipIdentity>>>`

This doesn't seem half bad as then:
1. `set_keypair()` becomes atomic, one just passes in a new `GossipIdentity`
2. Replay can hold a clone of `Arc<GossipIdentity>` and swap it out on a fork reset.
3. `GossipIdentity` internally holds the ugly logic of ensuring that its three fields hold the same pubkey (duplicated 3 times...)

It'll be a bunch of code churn to accomplish though though, wdyt?

Reply:
Yeah at this point I feel this might generate even more complexity.  How the stages that deal with the identity keypair are pretty simple, they can just refresh it at a time that's convenient for them.  There's no real need to make the swap fully atomic across the entire validator process